src/createHistoryMiddleware.js

```diff
@@ -0,0 +1,42 @@
+import ActionTypes from './ActionTypes';
+
+// This is a private action creator not intended to be called by users.
+function locationUpdated(location) {
+  return {
+    type: ActionTypes.LOCATION_UPDATED,
+    payload: location,
+  };
+}
+
+export default function createHistoryMiddleware(protocol) {
+  return function historyMiddleware({ dispatch }) {
+    const dispose = protocol.subscribe((location) => {
```

Comment:
This subscription is how we dispatch actions for e.g. the user hitting the back button (and also for the `GO` actions below, which rely on the browser's internal history tracking).